Add ruleset JSON files and accompanying README file

alarthast opened the pull request:
Related issue: #9. 
Previous iteration of this PR: #10, the comments on which have been addressed here. [New changes are shown here](https://github.com/opensafely-actions/.github/compare/dfde365f54c93bb635f9245d1f8fd6cf3f8ac799..d1459be97c5ea74e504d953815ee6787ad373797).

See the issue for context, and the commit message for implementation details.

The rulesets have been tested by trying to violate them on a separate user-owned public repo.
In each case of attempted rule violation (e.g., upon trying to delete the default branch), the attempt has been blocked in some way. Screenshots or copy-pasted terminal output of the blocked attempts are in the [issue comment](https://github.com/opensafely-actions/.github/issues/9#issuecomment-2741182773).

Note that the original issue comment wanted to create a rule to prevent users from pushing commits that include changes to files in the .github directory. This is not possible without a push ruleset, but push rulesets are not available for public repos. Since all our reusable action repos are public repos, it is not possible to implement such a rule. This has been noted in the issue. 

### Next steps
Merging this PR does not complete the issue, as the issue also requires the rulesets to be applied to all the OpenSAFELY-owned repositories in the opensafely-actions org. 
After this PR is merged, we should import the rulesets into the relevant repos to close the issue.

### Additional notes
Particularly, I'd like to request feedback on whether the rule to prevent force-pushing to a semver tag is worth keeping, as this was not mentioned in the original issue, but might be a good rule to enforce anyway:
https://github.com/opensafely-actions/.github/blob/27126de7b5eeb7f749f6df8990d4ee193f9340fa/rulesets/protect-semantically-versioned-tags.json#L18-L20

alarthast committed:
Add ruleset JSON files and accompanying README file

- We would like to use rulesets to better protect the default branch and semantically
versioned tags of reusable action repos.
- Add two JSON files which can be imported into GitHub's UI to create the desired rulesets.
- Also add a readme file to provide context and to link to relevant documentation. The
readme includes a reminder to manually add required status checks to the ruleset
that protects the default branch.

Co-authored-by: Iain Dillingham <iaindillingham@users.noreply.github.com>

iaindillingham commented on rulesets/README.md:
```diff
@@ -0,0 +1,50 @@
+This directory houses two JSON files that can be used to import rulesets into
+a repository in the `opensafely-actions` organization.
+We recommend you apply these rulesets to repositories housing reusable actions
+within the organization.
```
What about...

> This directory contains several [rulesets][1] for repositories in the opensafely-actions organisation.

We don't want to spend more time than necessary maintaining this README. We don't need to mention the number (two) or type (JSON) of thing (files). We don't need to make recommendations, as the purpose of these rulesets is clear.

[1]: https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets

alarthast replied:
I've changed to the suggested text

iaindillingham commented on rulesets/README.md:
```diff
@@ -0,0 +1,50 @@
+
+Please ensure that you add the required status checks for the repo. If you try to save the ruleset without
+adding any status checks, GitHub will raise an error (see the screenshot below). 
+![Screenshot showing a GitHub error stating that "Required status checks cannot be empty"](error-checks-required.png)
```
As a reader, I feel you're making me work: what's the minimum I need to know? I'd argue it's that the "Protect default branch" ruleset requires status checks to pass before a topic branch can be merged into the default branch; and that I have to specify the status checks myself in the UI.

Again, we don't want to spend more time than necessary maintaining this README, so I'd strongly suggest we avoid screenshots.

alarthast replied:
I've removed the screenshots and simplified the section.

iaindillingham commented on rulesets/README.md:
```diff
@@ -0,0 +1,50 @@
+- Prevent force-pushing
+
+Since semantically versioned tags are used to represent releases, the above protections are in place to
+ensure that the release history is preserved and that tags are not accidentally deleted or modified.
```
Let's reduce our maintenance burden! In this section, let's explain _why_ we've created each of these rulesets, but not describe _how_ we've implemented them.

You could link to the "[Available rules for rulesets][1]" page to help the reader interpret the rules.

[1]: https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/available-rules-for-rulesets

alarthast replied:
I've removed the bullet points for the specific rules, but left the rationale text in.

iaindillingham reviewed (commented):
Thank you for continuing to work on this! I've left a couple of inline comments about reducing the maintenance burden. I hope you don't find them nitpicky: you'll soon find that creating things is easy, but maintaining things is hard. Anything we can do to make maintaining them easier is worthwhile. Arguably, thinking like this means we're thinking like software _engineers_ rather than software _developers_.

I'd really encourage you to try to see a PR from the reviewer's perspective. The description has a lot of information: my first thought was, "What do you want me to do?" Some questions that popped into my head:

* Why do I need to see the commit message in the description?
* Why are there next steps? Is this work not finished?
* What's the problem with push rulesets?
* Where can I find the rule you want feedback on?

Please don't create another PR to address these issues. We're here now!

alarthast commented:
Thank you for taking a look at this! I've updated the PR comment to address the points you raised. Hopefully things are a bit more readable?

> Why do I need to see the commit message in the description?

You don't - I wanted to have this PR bear resemblance to its previous iteration, so I thought I'd keep the commit message in. If it leads to an overly-inflated PR comment, I've removed it.

iaindillingham commented on rulesets/README.md:
```diff
@@ -0,0 +1,28 @@
+## Importing a ruleset into a repository
+Follow [the instructions in the GitHub docs for importing a ruleset](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/managing-rulesets-for-a-repository#importing-a-ruleset).
+
+Note that the [Protect default branch ruleset](#protect-default-branch-protect-default-branchjson) requires status
```
The link is broken. But the document is so short that the link isn't necessary. Let's remove it.

```suggestion
Note that the Protect default branch ruleset requires status
```

iaindillingham reviewed (approved):
If you could remove the broken link, then that would be great! Thank you for your hard work. Next step: Let's apply these rulesets to all the OpenSAFELY-owned repositories in the opensafely-actions organisation. These are:

- [x] cohort-joiner
- [x] cohort-report
- [x] dataset-report
- [x] deciles-charts
- [x] demographic-standardisation
- [x] kaplan-meier-function
- [x] minimal-action (no applicable status checks, so the "Require status checks" rule is not applied)
- [x] project-dag
- [x] safetab
